at_secondary/at_secondary_server/lib/src/server/at_compaction_observer_impl.dart

```diff
@@ -0,0 +1,123 @@
+import 'dart:convert';
+
+import 'package:at_commons/at_commons.dart';
+import 'package:at_persistence_secondary_server/at_persistence_secondary_server.dart';
+import 'package:at_secondary/src/server/at_secondary_config.dart';
+import 'package:at_secondary/src/server/at_secondary_impl.dart';
+import 'package:at_utils/at_logger.dart';
+
+/// Implements the [AtCompactionObserver]
+class AtCompactionObserverImpl implements AtCompactionObserver {
+  late AtLogType atLogType;
+  int compactionStartTimeInEpoch = 0;
+  int sizeBeforeCompaction = 0;
+
+  var keyStore = SecondaryPersistenceStoreFactory.getInstance()
+      .getSecondaryPersistenceStore(
+          AtSecondaryServerImpl.getInstance().currentAtSign)!
+      .getSecondaryKeyStore();
+
+  final _logger = AtSignLogger('AtCompactionObserverImpl');
+
+  AtCompactionObserverImpl(this.atLogType) {
+    // Gets the key from secondary keystore.
+    // If key is found, retrieves the previousRun value and add to compactionStats;
+    // Else inserts a new key.
+    keyStore!.get(_getKey()).catchError((onError) {
+      _logger.finer('${_getKey()} not found in keystore. Inserting a new key.');
+    }).then((atData) => _insertInitialStats(atData));
+  }
+
+  void _insertInitialStats(atData) {
+    var compactionStats = CompactionStats()
+      ..nextRun = DateTime.now()
```

Comment:
Does the compaction job honor this nextRun time? i.e Is the computed next run date time and actual run date equal?

Reply:
For the nextrun, in the end method, we add the compaction frequency mins to previous run DateTime.